File: components/frontend/src/types/agui.ts
/**
 * AG-UI Protocol Types
 * TypeScript types for AG-UI events and messages.
 * 
 * Reference: https://docs.ag-ui.com/concepts/events
 * Reference: https://docs.ag-ui.com/concepts/messages
 */

// AG-UI Event Types
export const AGUIEventType = {
  // Lifecycle events
  RUN_STARTED: 'RUN_STARTED',
  RUN_FINISHED: 'RUN_FINISHED',
  RUN_ERROR: 'RUN_ERROR',

  // Step events
  STEP_STARTED: 'STEP_STARTED',
  STEP_FINISHED: 'STEP_FINISHED',

  // Text message events (streaming)
  TEXT_MESSAGE_START: 'TEXT_MESSAGE_START',
  TEXT_MESSAGE_CONTENT: 'TEXT_MESSAGE_CONTENT',
  TEXT_MESSAGE_END: 'TEXT_MESSAGE_END',

  // Tool call events (streaming)
  TOOL_CALL_START: 'TOOL_CALL_START',
  TOOL_CALL_ARGS: 'TOOL_CALL_ARGS',
  TOOL_CALL_END: 'TOOL_CALL_END',

  // State management events
  STATE_SNAPSHOT: 'STATE_SNAPSHOT',
  STATE_DELTA: 'STATE_DELTA',

  // Message snapshot for restore/reconnect
  MESSAGES_SNAPSHOT: 'MESSAGES_SNAPSHOT',

  // Activity events
  ACTIVITY_SNAPSHOT: 'ACTIVITY_SNAPSHOT',
  ACTIVITY_DELTA: 'ACTIVITY_DELTA',

  // Raw event
  RAW: 'RAW',
} as const

export type AGUIEventTypeValue = (typeof AGUIEventType)[keyof typeof AGUIEventType]

// AG-UI Message Roles
export const AGUIRole = {
  USER: 'user',
  ASSISTANT: 'assistant',
  SYSTEM: 'system',
  TOOL: 'tool',
  DEVELOPER: 'developer',
  ACTIVITY: 'activity',
} as const

export type AGUIRoleValue = (typeof AGUIRole)[keyof typeof AGUIRole]

// Base event structure
export type AGUIBaseEvent = {
  type: AGUIEventTypeValue
  threadId: string
  runId: string
  timestamp: string
  messageId?: string
  parentRunId?: string
}

// Run input/output types
export type AGUIRunAgentInput = {
  threadId?: string
  runId?: string
  parentRunId?: string
  messages?: AGUIMessage[]
  state?: Record<string, unknown>
  tools?: AGUIToolDefinition[]
  context?: Record<string, unknown>
}

export type AGUIRunAgentOutput = {
  threadId: string
  runId: string
  parentRunId?: string
  streamUrl?: string
}

// Message type
export type AGUIMessage = {
  id: string
  role: AGUIRoleValue
  content?: string
  toolCalls?: AGUIToolCall[]
  toolCallId?: string
  name?: string
  timestamp?: string
  metadata?: unknown
  parentToolUseId?: string  // For hierarchical tool calls (sub-agents)
  children?: AGUIMessage[]  // Nested tool calls under this tool
}

// Tool types
export type AGUIToolCall = {
  id: string
  name: string
  args: string
  type?: string
  parentToolUseId?: string  // For parent-child relationships (sub-agents)
  result?: string
  status?: 'pending' | 'running' | 'completed' | 'error'
  error?: string
  duration?: number
}

export type AGUIToolDefinition = {
  name: string
  description?: string
  parameters?: Record<string, unknown>
}

// Lifecycle events
export type AGUIRunStartedEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.RUN_STARTED
  input?: AGUIRunAgentInput
}

export type AGUIRunFinishedEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.RUN_FINISHED
  output?: unknown
}

export type AGUIRunErrorEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.RUN_ERROR
  error: string
  code?: string
  details?: string
}

// Step events
export type AGUIStepStartedEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.STEP_STARTED
  stepId: string
  stepName: string
}

export type AGUIStepFinishedEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.STEP_FINISHED
  stepId: string
  stepName: string
  duration?: number
}

// Text message events
export type AGUITextMessageStartEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.TEXT_MESSAGE_START
  role: AGUIRoleValue
}

export type AGUITextMessageContentEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.TEXT_MESSAGE_CONTENT
  delta: string
}

export type AGUITextMessageEndEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.TEXT_MESSAGE_END
}

// Tool call events
export type AGUIToolCallStartEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.TOOL_CALL_START
  toolCallId: string
  toolCallName: string
  parentMessageId?: string
  parentToolUseId?: string
}

export type AGUIToolCallArgsEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.TOOL_CALL_ARGS
  toolCallId: string
  delta: string
}

export type AGUIToolCallEndEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.TOOL_CALL_END
  toolCallId: string
  result?: string
  error?: string
  duration?: number
}

// State events
export type AGUIStateSnapshotEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.STATE_SNAPSHOT
  state: Record<string, unknown>
}

export type AGUIStateDeltaEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.STATE_DELTA
  delta: AGUIStatePatch[]
}

export type AGUIStatePatch = {
  op: 'add' | 'remove' | 'replace'
  path: string
  value?: unknown
}

// Message snapshot event
export type AGUIMessagesSnapshotEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.MESSAGES_SNAPSHOT
  messages: AGUIMessage[]
}

// Activity types
export type AGUIActivity = {
  id: string
  type: string
  title?: string
  status?: 'pending' | 'running' | 'completed' | 'error'
  progress?: number
  data?: Record<string, unknown>
}

export type AGUIActivitySnapshotEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.ACTIVITY_SNAPSHOT
  activities: AGUIActivity[]
}

export type AGUIActivityDeltaEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.ACTIVITY_DELTA
  delta: AGUIActivityPatch[]
}

export type AGUIActivityPatch = {
  op: 'add' | 'update' | 'remove'
  activity: AGUIActivity
}

// Raw event
export type AGUIRawEvent = AGUIBaseEvent & {
  type: typeof AGUIEventType.RAW
  data: unknown
}

// Union of all event types
export type AGUIEvent =
  | AGUIRunStartedEvent
  | AGUIRunFinishedEvent
  | AGUIRunErrorEvent
  | AGUIStepStartedEvent
  | AGUIStepFinishedEvent
  | AGUITextMessageStartEvent
  | AGUITextMessageContentEvent
  | AGUITextMessageEndEvent
  | AGUIToolCallStartEvent
  | AGUIToolCallArgsEvent
  | AGUIToolCallEndEvent
  | AGUIStateSnapshotEvent
  | AGUIStateDeltaEvent
  | AGUIMessagesSnapshotEvent
  | AGUIActivitySnapshotEvent
  | AGUIActivityDeltaEvent
  | AGUIRawEvent

// Run metadata type
export type AGUIRunMetadata = {
  threadId: string
  runId: string
  parentRunId?: string
  sessionName: string
  projectName: string
  startedAt: string
  finishedAt?: string
  status: 'running' | 'completed' | 'error'
  eventCount?: number
  restartCount?: number
}

// History response type
export type AGUIHistoryResponse = {
  threadId: string
  runId?: string
  messages: AGUIMessage[]
  runs: AGUIRunMetadata[]
}

// Runs response type
export type AGUIRunsResponse = {
  threadId: string
  runs: AGUIRunMetadata[]
}

// Pending tool call being streamed
export type PendingToolCall = {
  id: string
  name: string
  args: string
  parentToolUseId?: string
}

// Client state for AG-UI streaming
export type AGUIClientState = {
  threadId: string | null
  runId: string | null
  status: 'idle' | 'connecting' | 'connected' | 'error' | 'completed'
  messages: AGUIMessage[]
  state: Record<string, unknown>
  activities: AGUIActivity[]
  currentMessage: {
    id: string | null
    role: AGUIRoleValue | null
    content: string
  } | null
  // DEPRECATED: Use pendingToolCalls instead for parallel tool call support
  currentToolCall: {
    id: string | null
    name: string | null
    args: string
    parentToolUseId?: string
  } | null
  // Track ALL in-progress tool calls (supports parallel tool execution)
  pendingToolCalls: Map<string, PendingToolCall>
  // Track child tools that finished before their parent
  pendingChildren: Map<string, AGUIMessage[]>
  error: string | null
}

// Type guard functions
export function isRunStartedEvent(event: AGUIEvent): event is AGUIRunStartedEvent {
  return event.type === AGUIEventType.RUN_STARTED
}

export function isRunFinishedEvent(event: AGUIEvent): event is AGUIRunFinishedEvent {
  return event.type === AGUIEventType.RUN_FINISHED
}

export function isRunErrorEvent(event: AGUIEvent): event is AGUIRunErrorEvent {
  return event.type === AGUIEventType.RUN_ERROR
}

export function isTextMessageStartEvent(event: AGUIEvent): event is AGUITextMessageStartEvent {
  return event.type === AGUIEventType.TEXT_MESSAGE_START
}

export function isTextMessageContentEvent(event: AGUIEvent): event is AGUITextMessageContentEvent {
  return event.type === AGUIEventType.TEXT_MESSAGE_CONTENT
}

export function isTextMessageEndEvent(event: AGUIEvent): event is AGUITextMessageEndEvent {
  return event.type === AGUIEventType.TEXT_MESSAGE_END
}

export function isToolCallStartEvent(event: AGUIEvent): event is AGUIToolCallStartEvent {
  return event.type === AGUIEventType.TOOL_CALL_START
}

export function isToolCallEndEvent(event: AGUIEvent): event is AGUIToolCallEndEvent {
  return event.type === AGUIEventType.TOOL_CALL_END
}

export function isStateSnapshotEvent(event: AGUIEvent): event is AGUIStateSnapshotEvent {
  return event.type === AGUIEventType.STATE_SNAPSHOT
}

export function isMessagesSnapshotEvent(event: AGUIEvent): event is AGUIMessagesSnapshotEvent {
  return event.type === AGUIEventType.MESSAGES_SNAPSHOT
}

export function isActivitySnapshotEvent(event: AGUIEvent): event is AGUIActivitySnapshotEvent {
  return event.type === AGUIEventType.ACTIVITY_SNAPSHOT
}

File: components/backend/types/agui.go
// Package types defines AG-UI protocol types for event streaming.
// Reference: https://docs.ag-ui.com/concepts/events
package types

import "time"

// AG-UI Event Types as defined in the protocol specification
// See: https://docs.ag-ui.com/concepts/events
const (
	// Lifecycle events
	EventTypeRunStarted  = "RUN_STARTED"
	EventTypeRunFinished = "RUN_FINISHED"
	EventTypeRunError    = "RUN_ERROR"

	// Step events
	EventTypeStepStarted  = "STEP_STARTED"
	EventTypeStepFinished = "STEP_FINISHED"

	// Text message events (streaming)
	EventTypeTextMessageStart   = "TEXT_MESSAGE_START"
	EventTypeTextMessageContent = "TEXT_MESSAGE_CONTENT"
	EventTypeTextMessageEnd     = "TEXT_MESSAGE_END"

	// Tool call events (streaming)
	EventTypeToolCallStart = "TOOL_CALL_START"
	EventTypeToolCallArgs  = "TOOL_CALL_ARGS"
	EventTypeToolCallEnd   = "TOOL_CALL_END"

	// State management events
	EventTypeStateSnapshot = "STATE_SNAPSHOT"
	EventTypStateDelta     = "STATE_DELTA"

	// Message snapshot for restore/reconnect
	EventTypeMessagesSnapshot = "MESSAGES_SNAPSHOT"

	// Activity events (frontend-only durable UI)
	EventTypeActivitySnapshot = "ACTIVITY_SNAPSHOT"
	EventTypeActivityDelta    = "ACTIVITY_DELTA"

	// Raw event for pass-through
	EventTypeRaw = "RAW"
)

// AG-UI Message Roles
// See: https://docs.ag-ui.com/concepts/messages
const (
	RoleUser      = "user"
	RoleAssistant = "assistant"
	RoleSystem    = "system"
	RoleTool      = "tool"
	RoleDeveloper = "developer"
	RoleActivity  = "activity"
)

// BaseEvent is the common structure for all AG-UI events
// See: https://docs.ag-ui.com/concepts/events#baseeventproperties
type BaseEvent struct {
	Type      string `json:"type"`
	ThreadID  string `json:"threadId"`
	RunID     string `json:"runId"`
	Timestamp string `json:"timestamp"`
	// Optional fields
	MessageID   string `json:"messageId,omitempty"`
	ParentRunID string `json:"parentRunId,omitempty"`
}

// RunAgentInput is the input format for starting an AG-UI run
// See: https://docs.ag-ui.com/quickstart/introduction
type RunAgentInput struct {
	ThreadID    string                 `json:"threadId,omitempty"`
	RunID       string                 `json:"runId,omitempty"`
	ParentRunID string                 `json:"parentRunId,omitempty"`
	Messages    []Message              `json:"messages,omitempty"`
	State       map[string]interface{} `json:"state,omitempty"`
	Tools       []ToolDefinition       `json:"tools,omitempty"`
	Context     map[string]interface{} `json:"context,omitempty"`
}

// RunAgentOutput is the response after starting a run
type RunAgentOutput struct {
	ThreadID    string `json:"threadId"`
	RunID       string `json:"runId"`
	ParentRunID string `json:"parentRunId,omitempty"`
	StreamURL   string `json:"streamUrl,omitempty"`
}

// Message represents an AG-UI message in the conversation
// See: https://docs.ag-ui.com/concepts/messages
type Message struct {
	ID         string      `json:"id"`
	Role       string      `json:"role"`
	Content    string      `json:"content,omitempty"`
	ToolCalls  []ToolCall  `json:"toolCalls,omitempty"`
	ToolCallID string      `json:"toolCallId,omitempty"`
	Name       string      `json:"name,omitempty"`
	Timestamp  string      `json:"timestamp,omitempty"`
	Metadata   interface{} `json:"metadata,omitempty"`
}

// ToolCall represents a tool call made by the assistant
type ToolCall struct {
	ID              string `json:"id"`
	Name            string `json:"name"`
	Args            string `json:"args"`
	Type            string `json:"type,omitempty"`            // "function"
	ParentToolUseID string `json:"parentToolUseId,omitempty"` // For hierarchical nesting
	Result          string `json:"result,omitempty"`
	Status          string `json:"status,omitempty"` // "pending", "running", "completed", "error"
	Error           string `json:"error,omitempty"`
	Duration        int64  `json:"duration,omitempty"` // milliseconds
}

// ToolDefinition describes an available tool
type ToolDefinition struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description,omitempty"`
	Parameters  map[string]interface{} `json:"parameters,omitempty"`
}

// RunStartedEvent is emitted when a run begins
type RunStartedEvent struct {
	BaseEvent
	Input *RunAgentInput `json:"input,omitempty"`
}

// RunFinishedEvent is emitted when a run completes successfully
type RunFinishedEvent struct {
	BaseEvent
	Output interface{} `json:"output,omitempty"`
}

// RunErrorEvent is emitted when a run fails
type RunErrorEvent struct {
	BaseEvent
	Error   string `json:"error"`
	Code    string `json:"code,omitempty"`
	Details string `json:"details,omitempty"`
}

// StepStartedEvent marks the beginning of a processing step
type StepStartedEvent struct {
	BaseEvent
	StepID   string `json:"stepId"`
	StepName string `json:"stepName"`
}

// StepFinishedEvent marks the completion of a processing step
type StepFinishedEvent struct {
	BaseEvent
	StepID   string `json:"stepId"`
	StepName string `json:"stepName"`
	Duration int64  `json:"duration,omitempty"` // milliseconds
}

// TextMessageStartEvent begins a streaming text message
type TextMessageStartEvent struct {
	BaseEvent
	Role string `json:"role"`
}

// TextMessageContentEvent contains a chunk of text content
type TextMessageContentEvent struct {
	BaseEvent
	Delta string `json:"delta"`
}

// TextMessageEndEvent marks the end of a streaming text message
type TextMessageEndEvent struct {
	BaseEvent
}

// ToolCallStartEvent begins a streaming tool call
type ToolCallStartEvent struct {
	BaseEvent
	ToolCallID      string `json:"toolCallId"`
	ToolCallName    string `json:"toolCallName"`
	ParentMessageID string `json:"parentMessageId,omitempty"`
	ParentToolUseID string `json:"parentToolUseId,omitempty"`
}

// ToolCallArgsEvent contains a chunk of tool call arguments
type ToolCallArgsEvent struct {
	BaseEvent
	ToolCallID string `json:"toolCallId"`
	Delta      string `json:"delta"`
}

// ToolCallEndEvent marks the end of a streaming tool call
type ToolCallEndEvent struct {
	BaseEvent
	ToolCallID string `json:"toolCallId"`
	Result     string `json:"result,omitempty"`
	Error      string `json:"error,omitempty"`
	Duration   int64  `json:"duration,omitempty"` // milliseconds
}

// StateSnapshotEvent provides complete state for hydration
type StateSnapshotEvent struct {
	BaseEvent
	State map[string]interface{} `json:"state"`
}

// StateDeltaEvent provides incremental state updates
type StateDeltaEvent struct {
	BaseEvent
	Delta []StatePatch `json:"delta"`
}

// StatePatch represents a JSON Patch operation for state updates
type StatePatch struct {
	Op    string      `json:"op"`   // "add", "remove", "replace"
	Path  string      `json:"path"` // JSON Pointer
	Value interface{} `json:"value,omitempty"`
}

// MessagesSnapshotEvent provides complete message history for hydration
type MessagesSnapshotEvent struct {
	BaseEvent
	Messages []Message `json:"messages"`
}

// ActivitySnapshotEvent provides complete activity UI state
type ActivitySnapshotEvent struct {
	BaseEvent
	Activities []Activity `json:"activities"`
}

// ActivityDeltaEvent provides incremental activity updates
type ActivityDeltaEvent struct {
	BaseEvent
	Delta []ActivityPatch `json:"delta"`
}

// Activity represents a durable frontend UI element
type Activity struct {
	ID       string                 `json:"id"`
	Type     string                 `json:"type"`
	Title    string                 `json:"title,omitempty"`
	Status   string                 `json:"status,omitempty"` // "pending", "running", "completed", "error"
	Progress float64                `json:"progress,omitempty"`
	Data     map[string]interface{} `json:"data,omitempty"`
}

// ActivityPatch represents an update to an activity
type ActivityPatch struct {
	Op       string   `json:"op"` // "add", "update", "remove"
	Activity Activity `json:"activity"`
}

// RawEvent allows pass-through of arbitrary data
type RawEvent struct {
	BaseEvent
	Data interface{} `json:"data"`
}

// NewBaseEvent creates a new BaseEvent with current timestamp
func NewBaseEvent(eventType, threadID, runID string) BaseEvent {
	return BaseEvent{
		Type:      eventType,
		ThreadID:  threadID,
		RunID:     runID,
		Timestamp: time.Now().UTC().Format(time.RFC3339Nano),
	}
}

// WithMessageID adds a message ID to the event
func (e BaseEvent) WithMessageID(messageID string) BaseEvent {
	e.MessageID = messageID
	return e
}

// WithParentRunID adds a parent run ID to the event
func (e BaseEvent) WithParentRunID(parentRunID string) BaseEvent {
	e.ParentRunID = parentRunID
	return e
}

// AGUIEventLog represents the persisted event log structure
type AGUIEventLog struct {
	ThreadID    string      `json:"threadId"`
	RunID       string      `json:"runId"`
	ParentRunID string      `json:"parentRunId,omitempty"`
	Events      []BaseEvent `json:"events"`
	CreatedAt   string      `json:"createdAt"`
	UpdatedAt   string      `json:"updatedAt"`
}

// AGUIRunMetadata contains metadata about a run for indexing
type AGUIRunMetadata struct {
	ThreadID     string `json:"threadId"`
	RunID        string `json:"runId"`
	ParentRunID  string `json:"parentRunId,omitempty"`
	SessionName  string `json:"sessionName"`
	ProjectName  string `json:"projectName"`
	StartedAt    string `json:"startedAt"`
	FinishedAt   string `json:"finishedAt,omitempty"`
	Status       string `json:"status"` // "running", "completed", "error"
	EventCount   int    `json:"eventCount"`
	RestartCount int    `json:"restartCount,omitempty"`
}
